crates/filtering/src/madgwick.rs

```diff
@@ -0,0 +1,83 @@
+use std::time::Duration;
+
+use nalgebra::{matrix, vector, Quaternion, UnitQuaternion, Vector3};
+
+#[derive(Debug)]
+pub struct AccelerometerNormZero;
+
+/// Madgwick's filter for orientation estimation.
+///
+/// This implementation is based on the following:
+/// <https://github.com/jmagnuson/ahrs-rs>
+pub trait Madgwick {
+    fn update_with_imu(
+        &mut self,
+        gyroscope: Vector3<f32>,
+        accelerometer: Vector3<f32>,
+        filter_gain: f32,
+        sample_period: Duration,
+    ) -> Result<(), AccelerometerNormZero>;
+
+    fn update_with_gyroscope(&mut self, gyroscope: Vector3<f32>, sample_period: Duration);
+}
+
+impl Madgwick for UnitQuaternion<f32> {
+    #[allow(non_snake_case)]
+    fn update_with_imu(
+        &mut self,
+        gyroscope: Vector3<f32>,
+        accelerometer: Vector3<f32>,
+        filter_gain: f32,
+        sample_period: Duration,
+    ) -> Result<(), AccelerometerNormZero> {
+        let q = self.as_ref();
+
+        // Normalize accelerometer measurement
+        let Some(accel) = accelerometer.try_normalize(0.0) else {
+            return Err(AccelerometerNormZero);
+        };
+
+        // Gradient descent algorithm corrective step
+        #[rustfmt::skip]
+        let F = vector![
+            2.0 * (      q.i * q.k - q.w * q.j) - accel.x,
+            2.0 * (      q.w * q.i + q.j * q.k) - accel.y,
+            2.0 * (0.5 - q.i * q.i - q.j * q.j) - accel.z,
+            0.0
+        ];
```

Comment:
```suggestion
let reference = Quaternion::from_vector(Vector3::z());
let F = (q.conjugate() * reference * q).as_vector()  - accel.to_homogeneous()
```

Reply:
talked to @schmidma. Although this suggestion does not change the filter, we will implement these steps later, tracked by #1133